mirgecom/limiter.py

```diff
@@ -0,0 +1,78 @@
+""":mod:`mirgecom.limiter` is for limiters and limiter-related constructs.
+
+Field limiter functions
+^^^^^^^^^^^^^^^^^^^^^^^
+
+.. autofunction:: positivity_preserving_limiter
+
+
+Helper Functions
+^^^^^^^^^^^^^^^^
+
+.. autofunction:: cell_volume
+"""
+
+__copyright__ = """
+Copyright (C) 2022 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from grudge.discretization import DiscretizationCollection
+import grudge.op as op
+
+
+def cell_characteristic_size(actx, dcoll: DiscretizationCollection):
+    """Evaluate cell area or volume."""
+    zeros = actx.thaw(actx.freeze(dcoll.nodes()))[0]
+    return op.elementwise_integral(dcoll, zeros + 1.0)
+
+
+def positivity_preserving_limiter(dcoll: DiscretizationCollection, volume, field,
+                                  mmin=0.0, mmax=1.0):
+    """Implement the positivity-preserving limiter of Liu and Osher (1996)."""
```

Comment:
Document the roles of `volume` and `field`, as well as the return value. Ideally also type-annotate them.